In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as mlt


In [2]:
#there are two important factors to consider:
#format and file path.
# Format is the way data is encoded.
# We can usually tell different encoding schemes by looking at the ending of the file name.
# Some common encodings are csv, json, xlsx, hdf and so forth.
# The (file) path tells us where the data is stored.

In [3]:
# header ---> NONE
  # when set header is equal to None pandas automaticaly give the index to row and cloumn
  # header=[.......................]
  #df.columns=header  -------> its means assign column to data-set

In [4]:
# pandas Data type ::
    # object 
    # int64
    # float
    # datetime
# ----> Check data type --relative column
# #--------------> dtypes
# ---------------------------------------------------------------------------------------------
# ------------------------------------------------------------------------------------------------
# check the destribution of data in each column by ?describe? method--->
#    ---> It returns the number of terms in the column as
# "count", average column value as "mean",
                                # To enable a summary of all the columns, we could
                                # add an argument include = "all" inside
                                # We see that for object-type columns, a different set of 
                                # statistics is evaluated, like unique,
# top and frequency.
# "Unique" is the number of distinct objects in the column, "top" is the most frequently

In [5]:
df=pd.read_csv('car_data_set.csv',header=None)
df

0    1            2       3      4     5            6    7      8   \
0     3    ?  alfa-romero     gas    std   two  convertible  rwd  front   
1     3    ?  alfa-romero     gas    std   two  convertible  rwd  front   
2     1    ?  alfa-romero     gas    std   two    hatchback  rwd  front   
3     2  164         audi     gas    std  four        sedan  fwd  front   
4     2  164         audi     gas    std  four        sedan  4wd  front   
..   ..  ...          ...     ...    ...   ...          ...  ...    ...   
200  -1   95        volvo     gas    std  four        sedan  rwd  front   
201  -1   95        volvo     gas  turbo  four        sedan  rwd  front   
202  -1   95        volvo     gas    std  four        sedan  rwd  front   
203  -1   95        volvo  diesel  turbo  four        sedan  rwd  front   
204  -1   95        volvo     gas  turbo  four        sedan  rwd  front   

        9   ...   16    17    18    19    20   21    22  23  24     25  
0     88.6  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1     88.6  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2     94.5  ...  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3     99.8  ...  109  mpfi  3.19   3.4  10.0  102  5500  24  30  13950  
4     99.4  ...  136  mpfi  3.19   3.4   8.0  115  5500  18  22  17450  
..     ...  ...  ...   ...   ...   ...   ...  ...   ...  ..  ..    ...  
200  109.1  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
201  109.1  ...  141  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
202  109.1  ...  173  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
203  109.1  ...  145   idi  3.01   3.4  23.0  106  4800  26  27  22470  
204  109.1  ...  141  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[205 rows x 26 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       205 non-null    int64  
 1   1       205 non-null    object 
 2   2       205 non-null    object 
 3   3       205 non-null    object 
 4   4       205 non-null    object 
 5   5       205 non-null    object 
 6   6       205 non-null    object 
 7   7       205 non-null    object 
 8   8       205 non-null    object 
 9   9       205 non-null    float64
 10  10      205 non-null    float64
 11  11      205 non-null    float64
 12  12      205 non-null    float64
 13  13      205 non-null    int64  
 14  14      205 non-null    object 
 15  15      205 non-null    object 
 16  16      205 non-null    int64  
 17  17      205 non-null    object 
 18  18      205 non-null    object 
 19  19      205 non-null    object 
 20  20      205 non-null    float64
 21  21      205 non-null    object 
 22  22

##  Rename Specific Columns

In [7]:
# Method 1: Rename a specific column (recommended)
df.rename(columns={'old_column_name': 'new_column_name'}, inplace=True)

## Rename all Columns

In [8]:
# adder into it like column-wise
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
df.columns=headers
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

# How to Change the Data type
| errors value | Behavior                 |
| ------------ | ------------------------ |
| `'raise'`    | Throws error (default) ❌ |
| `'ignore'`   | Leaves column unchanged  |
| `'coerce'`   | Converts invalid → NaN ✅ |


In [10]:
df['bore']=pd.to_numeric(df['bore'],errors='coerce')
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

## Replace the Element ? with NaN

In [ ]:
df1=df.replace('?',np.nan)
df=df1.dropna(subset=['price'],axis=0)  #.........  subset also specify the column and operation perform on specific column
df.head(20) 
#----> rows deleted where price is nan  



symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           3               NaN  alfa-romero       gas        std   
2           1               NaN  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2               NaN         audi       gas        std   
6           1               158         audi       gas        std   
7           1               NaN         audi       gas        std   
8           1               158         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1               NaN          bmw       gas        std   
15          0               NaN          bmw       gas        std   
16          0               NaN          bmw       gas        std   
17          0               NaN          bmw       gas        std   
18          2               121    chevrolet       gas        std   
19          1                98    chevrolet       gas        std   
20          0                81    chevrolet       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
10          two        sedan          rwd           front       101.2  ...   
11         four        sedan          rwd           front       101.2  ...   
12          two        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       101.2  ...   
14         four        sedan          rwd           front       103.5  ...   
15         four        sedan          rwd           front       103.5  ...   
16          two        sedan          rwd           front       103.5  ...   
17         four        sedan          rwd           front       110.0  ...   
18          two    hatchback          fwd           front        88.4  ...   
19          two    hatchback          fwd           front        94.5  ...   
20         four        sedan          fwd           front        94.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19     3.4              10.0        102   
4           136         mpfi  3.19     3.4               8.0        115   
5           136         mpfi  3.19     3.4               8.5        110   
6           136         mpfi  3.19     3.4               8.5        110   
7           136         mpfi  3.19     3.4               8.5        110   
8           131         mpfi  3.13     3.4               8.3        140   
10          108         mpfi   3.5     2.8               8.8        101   
11          108         mpfi   3.5     2.8               8.8        101   
12          164

## Add value where NaN placed throught fillna()

In [ ]:
df['bore'].fillna(df['bore'].mean(),inplace=True)
# fill value the all coumn avg Values

## Specific Column to Perform Describe Function

In [11]:
## to Save the data Set
    #df.to_csv('automobile.csv',index=False)
    
#selected columns and describe
df[['length', 'compression-ratio']].describe()

length  compression-ratio
count  205.000000         205.000000
mean   174.049268          10.142537
std     12.337289           3.972040
min    141.100000           7.000000
25%    166.300000           8.600000
50%    173.200000           9.000000
75%    183.100000           9.400000
max    208.100000          23.000000

# Pre-Processing Data in Python

It is the process of converting or mapping data from one “raw” form into another

format to make it ready for further analysis.

Data pre-processing is also often called “data cleaning” or “data wrangling”, and there

In [ ]:
#-----------Identify and handle missing values

#-----------Format data into correct data types

#-----------Normalize data (centering / scaling)

#-----------Perform data binning

#-----------Convert categorical values into numeric variables


### What is Data Normalization?

**Data normalization means making numbers comparable** by putting them on a **similar scale**.

👉 Nothing fancy. Just **balancing numbers**.

---

### Real-life example (VERY important)

Imagine you’re comparing two students:

| Student | Height (cm) | Weight (kg) |
| ------- | ----------- | ----------- |
| A       | 170         | 65          |
| B       | 180         | 70          |

Now imagine a computer:

* Height values ≈ **170–180**
* Weight values ≈ **60–70**

Height numbers are **bigger**, so the computer thinks:

> “Height is more important than weight”

❌ **That’s wrong thinking**

---

### Why normalization is needed

Without normalization:

* Bigger numbers **dominate**
* Smaller numbers get **ignored**
* Model gives **wrong results**

Normalization fixes this.

---

### Type 1️⃣ Scaling (Min-Max Normalization)

#### Idea:

Bring values between **0 and 1**

#### Formula (don’t panic 😄):

```
(value − min) / (max − min)
```

#### Example:

Length values:

```
150, 160, 170
```

After scaling:

```
0.0, 0.5, 1.0
```

Now all values are **balanced**.

---

### Type 2️⃣ Centering (Standardization)

#### Idea:

Make:

* Average = **0**
* Spread = **1**

#### Formula:

```
(value − mean) / standard deviation
```

#### Example:

If mean = 170

| Original | After centering |
| -------- | --------------- |
| 160      | -1              |
| 170      | 0               |
| 180      | +1              |

Now:

* Negative = below average
* Positive = above average


### When do we use normalization?

✅ When:

* Features have **different units**
* Using ML models like:

  * KNN
  * Linear Regression
  * SVM
  * Neural Networks

❌ Not needed for:

* Decision Trees
* Random Forest







# Dealing with Missing Value in Python

Usually, missing value in dataset appears as “?”, “N/A”, 0 or just a blank cell.
these are the typical options you can consider:

• The first is to check if the person or group that collected the data can go back
and find what the actual value should be.

Another possibility is just to remove the data where that missing value is found.
• When you drop data, you can either drop the whole variable or just the single data
entry with the missing value.

If you don’t have a lot of observations with missing data, usually dropping the particular
entry is the best.

If you’re removing data, you want to look to do something that has the least amount
of impact.

Replacing data is better, since no data is wasted.
However, it is less accurate since we need to replace missing data with a guess of what
the data should be.

One standard replacement technique is to replace missing values by the average value of the

| Case                      | What to do        |
| ------------------------- | ----------------- |
| Numeric data missing      | Use mean / median |
| Categorical data missing  | Use mode          |
| Extra knowledge available | Fill logically    |


 ### df['fuel_type'].mode()

In [ ]:
#   df.dropna(subset=df['price'],axis=1,inplace= True) -------> its Working in data frames
#  df['price'].dropna(axis=1,inplace=True) -----> acutly working in Series

In [ ]:
### Replace Method

df.replace('!',df['price'].mean())

# replace the values by mean value of Price it Help to  Prevant the waste the Data
  


* **Check with the data collection source**
* **Drop the missing values**

  * Drop the variable
  * Drop the data entry
* **Replace the missing values**

  * Replace it with an average (of similar datapoints)
  * Replace it by frequency
  * Replace it based on other functions
* **Leave it as missing data**


# Data Formating by Python